libs/cosmos-sdk/baseapp/evm2cm.go

```diff
@@ -0,0 +1,118 @@
+package baseapp
+
+import (
+	"encoding/json"
+	"fmt"
+
+	sdk "github.com/okex/exchain/libs/cosmos-sdk/types"
+)
+
+var (
+	cmHandles          = make(map[string]*CMHandle)
+	evmResultConverter func(txHash, data []byte) ([]byte, error)
+	evmConvertJudge    func(msg sdk.Msg) ([]byte, bool)
+	evmParamParse      func(msg sdk.Msg) ([]byte, error)
+)
+
+type MsgWrapper struct {
+	Name string          `json:"type"`
+	Data json.RawMessage `json:"value"`
+}
+
+type CMHandle struct {
+	fn     func(data []byte, signers []sdk.AccAddress) (sdk.Msg, error)
+	height int64
+}
+
+func NewCMHandle(fn func(data []byte, signers []sdk.AccAddress) (sdk.Msg, error), height int64) *CMHandle {
+	return &CMHandle{
+		fn:     fn,
+		height: height,
+	}
+}
+
+func RegisterCmHandle(msgType string, create *CMHandle) {
+	if create == nil {
+		panic("Register CmHandle is nil")
+	}
+	if _, dup := cmHandles[msgType]; dup {
+		panic("Register CmHandle twice for same module and func " + msgType)
+	}
+	cmHandles[msgType] = create
+}
+
+func RegisterEvmResultConverter(create func(txHash, data []byte) ([]byte, error)) {
+	if create == nil {
+		panic("Register EvmResultConverter is nil")
+	}
+	evmResultConverter = create
+}
+
+func RegisterEvmParamParse(create func(msg sdk.Msg) ([]byte, error)) {
+	if create == nil {
+		panic("Register EvmParamParse is nil")
+	}
+	evmParamParse = create
+}
+
+func RegisterEvmConvertJudge(create func(msg sdk.Msg) ([]byte, bool)) {
+	if create == nil {
+		panic("Register EvmConvertJudge is nil")
+	}
+	evmConvertJudge = create
+}
+
+func ConvertMsg(msg sdk.Msg, height int64) (sdk.Msg, error) {
+	v, err := evmParamParse(msg)
+	if err != nil {
+		return nil, err
+	}
+	msgWrap, err := ParseMsgWrapper(v)
+	if err != nil {
+		return nil, err
+	}
+	if cmh, ok := cmHandles[msgWrap.Name]; ok && height >= cmh.height {
```

Comment:
does height >= cmh.height can use milestone in every msg with HigherThanVenusX func?

Reply:
> does height >= cmh.height can use milestone in every msg with HigherThanVenusX func?

two different way